### PR TITLE
Feature/housekeeping

### PR DESCRIPTION
**Pull request by s-block**

### CODE REVIEW

**Comment by s-block** [comment]
test this please

**Comment by s-block** [comment]
Can one of the admins verify this patch?

**Comment by s-block** [comment]
Can one of the admins verify this patch?

**Comment by s-block** [comment]
test this please

**Comment by s-block** [comment]
test this please

**Comment by s-block** [comment]
test this please

**Comment by s-block** [comment]
test this please

File: cla_backend/apps/cla_butler/qs_to_file.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        if hasattr(val, 'pk'):
+            val = val.pk
+        if val is None:
+            val = ''
+        try:
+            return unicode(val).encode('utf-8')
+        except UnicodeDecodeError:
+            return val
+
+    def dump_to_csv(self, qs):
```
Comment (andyhd) [inline cla_backend/apps/cla_butler/qs_to_file.py]: The three methods, `dump_to_csv`, `dump_to_yaml` and `dump_values` (which should be called `dump_to_json`), are all very similar and suggest a couple of refactorings.
1. Refactor `dump_to_csv` to extract the CSV writing code into a class or function that allows a similar pattern to the other methods, eg:
```python
with open(file_path, write_mode) as csvfile:
    csv.write(serialize_to_csv(qs))
```
2. Refactor the three methods into a single one which takes the serialization format/serialization function as a parameter, eg:
```python
def dump(self, qs, serialize_fn):
    ...
    with open(file_path, write_mode) as dump_file:
        dump_file.write(serialize_fn(qs))
    ...
```

File: cla_backend/apps/cla_butler/qs_to_file.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            writer = csv.writer(csvfile, quoting=csv.QUOTE_ALL)
+            if write_mode == WRITE_MODE:
+                writer.writerow(field_names)
+            for instance in qs.iterator():
+                writer.writerow(
+                    [self.get_value(instance, f) for f in
+                     qs.model._meta.fields])
+            csvfile.close()
+
+    def dump_to_yml(self, qs):
```
Comment (andyhd) [inline cla_backend/apps/cla_butler/qs_to_file.py]: Should be `dump_to_yaml`

File: cla_backend/apps/cla_butler/tasks.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
+# -*- coding: utf-8 -*-
+import datetime
+import os
+import time
+
+from celery import Task
+import logging
+
+from dateutil.relativedelta import relativedelta
```
Comment (andyhd) [inline cla_backend/apps/cla_butler/tasks.py]: PEP8 import grouping and ordering

File: cla_backend/apps/legalaid/models.py [hunk 4]
```diff
@@ -186,9 +186,13 @@ def add_string(s1, s2):
 
         for phone in [self.home_phone, self.mobile_phone]:
             if phone:
+                try:
+                    phone = unicode(phone)
+                except UnicodeDecodeError:
+                    pass
```
Comment (andyhd) [inline cla_backend/apps/legalaid/models.py]: Error not handled?

File: cla_backend/apps/cla_butler/tasks.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+            case__isnull=True,
+            modified__lte=yesterday,
+        )
+        pks = map(str, ecs.values_list('pk', flat=True))
+        self.cleanup_model_from_ec(pks, Property)
+        self._delete_objects(ecs)
+
+    def cleanup_model_from_ec(self, pks, model, attr='eligibility_check_id'):
+        qs = model.objects.filter(
+            **{'%s__in' % attr: pks}
```
Comment (andyhd) [inline cla_backend/apps/cla_butler/tasks.py]: This is hard to read. Precalculate the key in a variable, eg:
```python
attr_in = '{attribute}__in'.format(attribute=attr)
criteria = {attr_in: pks}
qs = model.objects.filter(**criteria)
```

File: cla_backend/apps/cla_butler/tasks.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        )
+        self._delete_objects(diags)
+
+    def cleanup_eligibility_check(self):
+        yesterday = self.now - datetime.timedelta(days=10)
+        ecs = EligibilityCheck.objects.filter(
+            case__isnull=True,
+            modified__lte=yesterday,
+        )
+        pks = map(str, ecs.values_list('pk', flat=True))
```
Comment (andyhd) [inline cla_backend/apps/cla_butler/tasks.py]: Prefer a list comprehension to `map`

File: cla_backend/apps/cla_butler/tasks.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        self._delete_objects(from_cases)
+        self._delete_objects(cases)
+
+    def cleanup_model_from_case(self, pks, model, attr='case_id', case_log_attr=None):
+        qs = model.objects.filter(
+            **{'%s__in' % attr: pks}
+        )
+        if case_log_attr:
+            logs = Log.objects.filter(
+                **{'%s__in' % case_log_attr: map(str, qs.values_list('pk', flat=True))}
```
Comment (andyhd) [inline cla_backend/apps/cla_butler/tasks.py]: There's a lot going on in this line. Should be split out for readability.

File: cla_backend/apps/cla_butler/tasks.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+            name=name))
+
+        logger.info('Deleting {count} {name} objects'.format(
+            count=qs.count(),
+            name=name))
+        self.filewriter.dump(qs)
+        logger.info('Starting delete of %s' % name)
+        start = time.time()
+        qs._raw_delete(qs.db)
+        logger.info('Time to delete %s: %s' % (name, time.time() - start))
```
Comment (andyhd) [inline cla_backend/apps/cla_butler/tasks.py]: Avoid mixing up `%` string interpolation and `format` - just use `format`

File: cla_backend/apps/cla_butler/tasks.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+        self.cleanup_sessions()
+
+    def _setup(self):
+        self.now = timezone.now()
+        path = os.path.join(settings.TEMP_DIR, self.now.strftime('%Y%m%d'))
+        self.filewriter = QuerysetToFile(path)
+
+    def _delete_logs(self, qs):
+        ct = ContentType.objects.get_for_model(qs.model)
+        pks = map(str, qs.values_list('pk', flat=True))
```
Comment (andyhd) [inline cla_backend/apps/cla_butler/tasks.py]: A list comprehension would be more readable here.

Comment (s-block) [inline cla_backend/apps/cla_butler/tasks.py]: you have changed

File: cla_backend/apps/cla_butler/qs_to_file.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+        elif val in field.empty_values:
+            val = None
+        elif isinstance(field, MoneyIntervalField):
+            val = MoneyInterval.from_dict(json.loads(val))
+        elif isinstance(field, JSONField):
+            val = json.dumps(val)
+        elif RE_DATE.match(val):
+            val = datetime.strptime(val, '%Y-%m-%d').date()
+        elif RE_DATETIME.match(val):
+            val = parse_datetime(val)
```
Comment (andyhd) [inline cla_backend/apps/cla_butler/qs_to_file.py]: Are regexes the best way to do this? Would [dateutil](https://dateutil.readthedocs.io/en/stable/parser.html) help here?

Comment (s-block) [inline cla_backend/apps/cla_butler/qs_to_file.py]: regex was faster that doing try/except converting in to date/datetime object

File: cla_backend/apps/legalaid/models.py (hunk 4 above)
Comment (s-block) [inline cla_backend/apps/legalaid/models.py]: don't want to do anything in case of error

File: cla_backend/apps/cla_butler/__init__.py [hunk 11]
```diff
@@ -0,0 +1 @@
+# -*- coding: utf-8 -*-
```
Comment (s-block) [inline cla_backend/apps/cla_butler/__init__.py]: My editor does this by default. I don't mind this in there but I can remove it if you want

Comment (andyhd) [inline cla_backend/apps/cla_butler/__init__.py]: Nope, just sayin

File: cla_backend/apps/cla_butler/qs_to_file.py (hunk 10 above)
Comment (andyhd) [inline cla_backend/apps/cla_butler/qs_to_file.py]: ok

**Comment by andyhd** [comment]
👍

**Pull request merged**